### ISSUE TITLE
Restore with locked-mode and no lockfile should fail

### ISSUE THREAD

**Issue by jvmdc** [NuGet/Home#13175]
### NuGet Product Used

dotnet.exe

### Product Version

8.0.100

### Impact

It bothers me. A fix would be nice

### Repro Steps & Context

1. Create a project without a lock-file, but with some dependency
2. Run dotnet restore --locked-mode
3. See that the project restore succeeds

My expectation would be that if I am attempting to force the project to restore using a locked mode, it would fail if no such lock is available. The project *will* fail if the file is present, but out of date or just plain empty.

Context: Building a centralized CI/CD setup in which I want to enforce the use of lock files to ensure reproducible builds. It seems there is no command that ensures a lock file is actually present built-in to the client. 

A workaround which I am currently using is to simply check for the presence of the file using a shell script wherever I detect a project-file. Built-in support in Nuget would be nice instead though, to make sure I do not miss corner-cases (e.g. not detecting a project file, not detecting a lock file whose location was changed through the `NuGetLockFilePath` parameter in the project file)

**Comment by kartheekp-ms** [comment]
I am able to reproduce the issue by following below steps.

- Created a new sample solution
- Created a new class library project in the solution
- Executed "dotnet add repro\repro.csproj package NuGet.Common" command to add a package reference to NuGet.Common for the class library project.
- Executed `dotnet restore --locked-mode` which succeeded despite the lack of lock file

```
{user-name} > ~\source\repos\test-13175 > dir

    Directory: C:\Users\{user-name}\source\repos\test-13175


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         1/23/2024   2:58 PM                repro
-a----         1/23/2024   2:58 PM            991 test-13175.sln


{user-name}  ~\source\repos\test-13175 ❯ dotnet restore
  Determining projects to restore...
  Restored C:\Users\{user-name}\source\repos\test-13175\repro\repro.csproj (in 277 ms).

{user-name} > ~\source\repos\test-13175 ❯ dotnet restore .\repro\repro.csproj --locked-mode
  Determining projects to restore...
  All projects are up-to-date for restore.
{user-name}  ~\source\repos\test-13175   
```

**Comment by jvmdc** [comment]
Just came back to this - I was wondering what "Quality Week" means as a label?

**Comment by to5ta** [comment]
I'd say its not only a nice requirement but a **must-have** for a professional use in CI/CD setups. Because, as already stated:

**How would you be sure to the CI/CD pipeline actually uses lock files?** (And does not simply _generate_ any missing lock files _silently_ instead of using it as _input_)

Impact imho is that you cannot fully trust LockedMode in CI but have to take extra measures as already described. 

PS: The issue was already reported back in 2019, see https://github.com/NuGet/Home/issues/8338

**Comment by jvmdc** [comment]
> PS: The issue was already reported back in 2019, see #8338

Yeah, I actually wrote a comment on that issue, waited a week with no reply and opened this one - Just forgot to link back to it, thanks for referencing it, so it's visible :)

**Comment by jvmdc** [comment]
@nkolev92 - I see that the tag "Quality Week" was added back in February. I haven't been able to find any documentation describing what Quality Week means, can you elaborate on this?

**Comment by nkolev92** [comment]
@jvmdc

The important label is priority label, which in this case is `Priority:2`, https://github.com/NuGet/Home/blob/dev/Issue-Triage-Policy.md. 

The Quality Week label subclasses backlog issues to aid our triage conversations.

**Comment by jvmdc** [comment]
@nkolev92 - Is the project open to us creating a Pull Request related to this issue?

Our proposal would be an implementation in which you can set `-LockedMode` to either `strict` or `lax`. `lax` would be the current (and default) behaviour if flag is specified without args (or LockedMode property set to true), whereas `strict` mode would fail restoration if a lock file was missing. This would require LockedMode to change type from boolean to a string/enum, so would be a breaking change for anyone using the public API - Would you prefer a new flag/Property instead?

**Comment by nkolev92** [comment]
If we're adding a new option, I think we'd need a design for that. https://github.com/NuGet/Home/blob/dev/meta/README.md. 

Is the purpose of making locked mode take options to avoid a potential breaking change? 
The one consideration is options like this often apply to *all* projects, but sometimes people have lock files only for certain projects. 

Final thing is consider that we'd need to replicate any option as a msbuild property (it's how many people set their values), so we'd probably need to work around that.

**Comment by jvmdc** [comment]
@nkolev92 
> Is the purpose of making locked mode take options to avoid a potential breaking change? 

Quite the opposite actually, as making it take options will break the public API of Nuget.ProjectModel (in which RestoreLockedMode is a boolean). The intention behind having it take options was to avoid clattering the CLI with extra flags that are situational, e.g. a new flag for this would only have relevance if `RestoreLockedMode` was set to true anyway.

> The one consideration is options like this often apply to all projects, but sometimes people have lock files only for certain projects.
The idea is that it would work similar to the existing behaviour where restore will fail if a lock-file is present but RestorePackagesWithLockFile is not enabled - Here is a (non-complete) example of a potential implementation: https://github.com/NuGet/NuGet.Client/compare/dev...jvmdc:NuGet.Client:patch-1

Does it make sense? If so, I'll get around to writing a proposal :)

**Comment by to5ta** [comment]
> If we're adding a new option, I think we'd need a design for that. https://github.com/NuGet/Home/blob/dev/meta/README.md. 
> 
> Is the purpose of making locked mode take options to avoid a potential breaking change? 
> The one consideration is options like this often apply to *all* projects, but sometimes people have lock files only for certain projects. 
> 
> Final thing is consider that we'd need to replicate any option as a msbuild property (it's how many people set their values), so we'd probably need to work around that. 
> 
> 

You can easily opt out restore with lock files per project e.g. by overwriting the  RestoreLockedMode property normally defined in a Directory.Build.props

**Comment by nkolev92** [comment]
I think seeing a design would be great. 

My recommendation would be to focus on the msbuild driven solution as well.

**Comment by zivkan** [comment]
> You can easily opt out restore with lock files per project e.g. by overwriting the RestoreLockedMode property normally defined in a Directory.Build.props

The point that @nkolev92 was making is that properties set on the command line are what MSBuild call global properties, and global properties cannot be overridden in props, targets, or project files.

> Our proposal would be an implementation in which you can set -LockedMode to either strict or lax. 

I don't believe that System.CommandLine supports optional arguments for options (i've been investigating exactly this for a new `dotnet package update` option). So, given `dotnet restore --locked-mode` is likely to be common in CI scripts, changing it to take a required parameter will probably break a lot of CI builds.

I agree that we should get a design, and whoever writes the design needs to take into consideration what is feasible within the constraints of System.CommandLine (since the dotnet CLI uses it for command line parsing), and MSBuild, and not introduce breaking changes.

Public APIs are less important, they can be broken if they need to be, since few people use Nuget.* packages in their own projects.  But it's too early to be thinking about NuGet's public APIs. We need a proposal for how to introduce the feature request without making customers angry with a breaking change.

**Comment by Frulfump** [comment]
From the docs https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-restore#options
```text
--locked-mode

Don't allow updating project lock file
```
Which I guess it can't do if there's not even a lock file present. I think people intuitively would expect the lock file MUST be present in `--locked-mode` (which is why we have this issue). But according to the docs it works as stated even tough it's unintuitive, would be good to get the docs amended to clarify.

I think adding another option would be the simplest solution and it avoids the breaking for people who are only using `--locked-mode` today
something like `--require-lock-file`

>The one consideration is options like this often apply to _all_ projects, but sometimes people have lock files only for certain projects.

Yes I have seen this in the wild and I think there has been some recommendations about which projects you add the lock file to? I think it's recommended to just add the lock file to top-level projects (EDIT mentioned here https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#make-lock-file-part-of-your-source-repository
>If you are building an application, an executable and the project in question is at the start of the dependency chain then do check in the lock file to the source code repository so that NuGet can make use of it during restore.
>
>However, if your project is a library project that you do not ship or a common code project on which other projects depend upon, you should not check in the lock file as part of your source code. There is no harm in keeping the lock file but the locked package dependencies for the common code project may not be used, as listed in the lock file, during the restore/build of a project that depends on this common-code project.).
>
 Ideally one top-level lock file would be present as described in #12409 that would remove this concern. I would expect this option to be applied for the projects in the graph where it makes sense i.e if Project A depends on Project B and Project A has `<RestorePackagesWithLockFile>true</ <RestorePackagesWithLockFile>` and I run restore on Project A I wouldn't expect this new option to trigger a warning for Project B.

But personally I would enforce lock-files in all projects as a policy if that was a requirement to get an expected failure when it's missing as I find it that important to have a reproducible restore (and I understand how the restore works in relation to https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#make-lock-file-part-of-your-source-repository).

(The docs for `use-lock-file` say this
```text
--use-lock-file

Enables project lock file to be generated and used with restore.
```
But isn't a requirement to get it to restore with a lockfile (EDIT: see docs excerpt below) as long as your project is using ` <RestorePackagesWithLockFile>true</ <RestorePackagesWithLockFile>` I don't think `--use-lock-file` should be a requirement to use lock files so should the docs for that be updated to include some more info? (People in general don't know or understand how MSBuild ties in to all of this.)

https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#enabling-the-lock-file
>Once a project has packages.lock.json file in its root directory, the lock file is always used with restore even if the property RestorePackagesWithLockFile is not set. So another way to opt-in to this feature is to create a dummy blank packages.lock.json file in the project's root directory.)
